### PR TITLE
Mkl bf16

### PR DESCRIPTION
**Pull request by mayurpandeyin**
# Existing Sample Changes
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes Issue# 

## External Dependencies

List any external dependencies created as a result of this change.

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Implement fixes for ONSAM Jiras

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode
- [ ] When compiling the compliler flag "-Wall -Wformat-security -Werror=format-security" was used

**_Delete this line and everything below if this is not a PR for a new code sample_**

**_Delete this line and all above it if this PR is for a new code sample_**
# Adding a New Sample(s)

## Description

Please include a description of the sample

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [ ] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [ ] Adhere to readme template 
- [ ] Enforce format via clang-format config file
- [ ] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

File: Libraries/oneMKL/matrix_mul_mkl/matrix_mul_mkl_bf16.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            catch (sycl::exception const& e) {
+                std::cout << "Caught asynchronous SYCL exception:\n"
+                    << e.what() << std::endl;
+            }
+            catch (std::exception const& e) {
+                std::cout << "Caught asynchronous STL exception:\n"
+                    << e.what() << std::endl;
+            }
+        }
+    };
```
Comment (petercad) [inline Libraries/oneMKL/matrix_mul_mkl/matrix_mul_mkl_bf16.cpp:45]: I'm not aware of any circumstances where DPC++ will throw asynchronous exceptions, so I wouldn't complicate this sample with an asynchronous exception handler.

File: Libraries/oneMKL/matrix_mul_mkl/matrix_mul_mkl_bf16.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+int64_t sizea = ldA * k, sizeb = ldB * n, sizec = ldC * n;
+double alpha = 1.0, beta = 0.0;
+// Allocate matrices
+std::vector<bfloat16> A(sizea);
+std::vector<bfloat16> B(sizeb);
+std::vector<bfloat16> C(sizec);
+// Initialize matrices […]
+std::fill(A.begin(), A.end(), 0);
+std::fill(B.begin(), B.end(), 0);
+std::fill(C.begin(), C.end(), 0);
```
Comment (petercad) [inline Libraries/oneMKL/matrix_mul_mkl/matrix_mul_mkl_bf16.cpp:25]: Please clean up the indenting/formatting here (formatting should match matrix_mul_mkl.cpp).

File: Libraries/oneMKL/matrix_mul_mkl/matrix_mul_mkl_bf16.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    // create sycl buffers of matrix data for offloading between device and host
+    sycl::buffer<bfloat16, 1> A_buffer(A.data(), A.size());
+    sycl::buffer<bfloat16, 1> B_buffer(B.data(), B.size());
+    sycl::buffer<bfloat16, 1> C_buffer(C.data(), C.size());
+    // add oneapi::mkl::blas::gemm to execution queue and catch any synchronous exceptions
+    try {
+        using oneapi::mkl::blas::gemm;
+        using oneapi::mkl::transpose;
+        gemm(my_queue, transpose::nontrans, transpose::nontrans, m, n, k, alpha, A_buffer, ldA, B_buffer,
+           ldB, beta, C_buffer, ldC);
```
Comment (petercad) [inline Libraries/oneMKL/matrix_mul_mkl/matrix_mul_mkl_bf16.cpp:57]: `my_queue.wait_and_throw()` should be placed here inside the `try` so any execution-related exceptions can be handled appropriately.

File: Libraries/oneMKL/matrix_mul_mkl/GNUmakefile
Comment (petercad) [inline Libraries/oneMKL/matrix_mul_mkl/GNUmakefile:1]: The Windows nmake Makefile will also need updating.

**Review by petercad — CHANGES_REQUESTED** [review]
You'll also want to update the README for the sample, providing some context for the new source file as well as a high-level description of what it's doing. Keep in mind that many people may not be familiar with the bfloat16 data type.

**Comment by jimmytwei** [comment]
@mayurpandeyin Can you make the updates based on @petercad's comments? Also, please resolve the conflict in 
Libraries/oneMKL/matrix_mul_mkl/makefile

Once you have done that, I will try build and run the sample and approve.

**Comment by jimmytwei** [comment]
@mayurpandeyin There's a conflict with Libraries/oneMKL/matrix_mul_mkl/makefile. Can you resolve that?
Also, dpcpp is deprecated. I think you will want to use icx-cl in the Makefile.

**Comment by jimmytwei** [comment]
@mayurpandeyin I'm going to close this issue out. We've made some changes to the existing matrix_mul_mkl sample. There was a request to modify the sample to have both a fp32 and fp64 variant of matrix_mul_mkl. If you want to add an example for bf16, I would suggest you to submit a new PR based off of the ``development`` branch.

**Pull request closed without merging**